kube/src/client/mod.rs

```diff
@@ -77,103 +76,95 @@ impl Client {
         Self::try_from(client_config)
     }
 
-    async fn send(&self, request: http::Request<Vec<u8>>) -> Result<reqwest::Response> {
-        let (parts, body) = request.into_parts();
-        let pandq = parts.uri.path_and_query().expect("valid path+query from kube");
-        let uri_str = finalize_url(&self.cluster_url, &pandq);
+    async fn send(&mut self, request: Request<Body>) -> Result<Response<Body>> {
+        let (mut parts, body) = request.into_parts();
         //trace!("Sending request => method = {} uri = {}", parts.method, &uri_str);
-
-        let mut headers = parts.headers;
+        let mut headers = self.headers.clone();
+        headers.extend(parts.headers.clone().into_iter());
         // If we have auth headers set, make sure they are updated and attached to the request
-        if let Some(auth_header) = self.config.get_auth_header().await? {
-            headers.insert(reqwest::header::AUTHORIZATION, auth_header);
+        if let Some(auth_header) = self.auth_header.to_header().await? {
+            headers.insert(http::header::AUTHORIZATION, auth_header);
         }
+        parts.headers = headers;
 
         let request = match parts.method {
             http::Method::GET
             | http::Method::POST
             | http::Method::DELETE
             | http::Method::PUT
-            | http::Method::PATCH => self.inner.request(parts.method, &uri_str),
+            | http::Method::PATCH => Request::from_parts(parts, body),
             other => return Err(Error::InvalidMethod(other.to_string())),
         };
 
-        let req = request.headers(headers).body(body).build()?;
-        let res = self.inner.execute(req).await?;
+        let res = self
+            .inner
+            .ready_and()
+            .await
+            .map_err(Error::Service)?
+            .call(request)
```

Comment:
> I see there's `mut` on all the `Api` methods now, presumably bubbling up from `Client::inner` AFAIKT. This might be worth hiding with an interior mutability wrapper.

Yeah, it's because of this `Service::call` taking a mutable reference. It was necessary to get started with `Service`.

I still need to figure out how `auth_header` is handled (async and can fail), but `Client::inner` is now a [`Buffer`](https://docs.rs/tower/0.4.4/tower/buffer/struct.Buffer.html), so we can revert by cloning `inner` in `send`.